# Introduction

This notebook introduces the PRIMAP-hist datasets exploration whose the links are available on our **[github repository](https://github.com/OpenGeoScales/ogs-data-exploration/tree/main/data/ghg-emissions/primap-hist)**. 

These datasets combine several published datasets and are provided by the Potsdam Institute for Climate Impact Research (PIK).

They provide a set of emissions pathways for each country and Kyoto Gas (GHG) from 1850 to 2018.

Here are the differents datasets from the last version (2.2) of February 2021:
- PRIMAP-hist_v2.2_19-Jan-2021.csv contains the numerical extrapolations of all time series to 2018
- PRIMAP-hist_v2.2_no_extrapolation_19-Jan-2021.csv has no numerical extrapolations of missing values and does not include the country groups "EARTH", "ANNEXI", "NONANNEXI", "AOSIS", "BASIC", "EU28", "LDC", "UMBRELLA"

Given that the first dataset contains no missing values thanks to the numerical extrapolations. The exploration below is focused on this dataset.

# Libraries importations and datasets loading

In [1]:
import os

import numpy as np
import pandas as pd

%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
folder = "../../../../PRIMAP-hist_v2.2/"
dataset_name = "PRIMAP-hist_v2.2_19-Jan-2021.csv"
primap_hist_data = pd.read_csv(folder+dataset_name)

In [5]:
primap_hist_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28622 entries, 0 to 28621
Columns: 174 entries, scenario to 2018
dtypes: float64(169), object(5)
memory usage: 38.0+ MB


In [6]:
primap_hist_data.head()

scenario country category entity unit      1850      1851      1852  \
0   HISTCR     ABW    IPC1A    CH4   Gg  0.000153  0.000158  0.000164   
1   HISTCR     AFG    IPC1A    CH4   Gg  0.023700  0.023800  0.023900   
2   HISTCR     AGO    IPC1A    CH4   Gg  1.750000  1.770000  1.800000   
3   HISTCR     AIA    IPC1A    CH4   Gg  0.000010  0.000010  0.000010   
4   HISTCR     ALB    IPC1A    CH4   Gg  0.060200  0.060600  0.061500   

       1853      1854  ...      2009     2010      2011      2012      2013  \
0  0.000169  0.000174  ...   0.03550   0.0385   0.03170   0.05170   0.06840   
1  0.024100  0.024200  ...   4.07000   4.4900   5.22000   7.82000  11.70000   
2  1.820000  1.840000  ...  57.70000  59.1000  60.50000  61.70000  63.00000   
3  0.000010  0.000010  ...   0.00205   0.0024   0.00256   0.00256   0.00275   
4  0.063100  0.065200  ...   5.14000   5.0400   4.98000   4.87000   4.79000   

       2014      2015      2016      2017     2018  
0   0.05940   0.05950   0.05390   0.05520   0.0565  
1   5.42000   5.38000   7.59000   7.96000   8.3300  
2  64.30000  65.60000  66.90000  68.30000  69.6000  
3   0.00275   0.00276   0.00286   0.00298   0.0031  
4   4.66000   4.89000   5.00000   5.02000   5.0400  

[5 rows x 174 columns]

# Variables and their description

The variables description that follows can be found **[here](https://zenodo.org/record/4479172#data-format-description-columns)**.

## Scenario

- HISTCR: in this scenario country-reported data (CRF, BUR, UNFCCC) is prioritized over third-party data (CDIAC, FAO, Andrew, EDGAR, BP).
- HISTTP: in this scenario third-party data (CDIAC, FAO, Andrew, EDGAR, BP) is prioritized over country-reported data (CRF, BUR, UNFCCC).

CRF, BUR, UNFCCC, CDIAC, FAO, Andrew, EDGAR and BP represent the data sources of the PIK dataset.

*How are the data prioritized over the others?*

In the article, it says that the highest priority datasets are the country reported data (UNFCCC). But if the data aren't available or don't meet the minimal requirements, then it uses country reported data provided by third-party sources such as reseach institutions or international organizations.
I haven't found how the scenarios are built and how the priorization is made.

*Does the priorization mean that the data are used from one category and not the other?*
*For all the years of a time serie? Or just some of them?*

I don't know. **For the two last questions, some exploration can be made to answer it.**

In [13]:
primap_hist_data.scenario.value_counts()

HISTTP    14311
HISTCR    14311
Name: scenario, dtype: int64

There are the same numbers of emission pathways (or time series) in each scenario.

## Country

207 ISO 3166 three-letter country codes or 8 custom codes for groups

In [14]:
primap_hist_data.country.unique()

array(['ABW', 'AFG', 'AGO', 'AIA', 'ALB', 'AND', 'ANNEXI', 'ANT', 'AOSIS',
       'ARE', 'ARG', 'ARM', 'ATG', 'AUS', 'AUT', 'AZE', 'BASIC', 'BDI',
       'BEL', 'BEN', 'BFA', 'BGD', 'BGR', 'BHR', 'BHS', 'BIH', 'BLR',
       'BLZ', 'BOL', 'BRA', 'BRB', 'BRN', 'BTN', 'BWA', 'CAF', 'CAN',
       'CHE', 'CHL', 'CHN', 'CIV', 'CMR', 'COD', 'COG', 'COK', 'COL',
       'COM', 'CPV', 'CRI', 'CUB', 'CYP', 'CZE', 'DEU', 'DJI', 'DMA',
       'DNK', 'DOM', 'DZA', 'EARTH', 'ECU', 'EGY', 'ERI', 'ESP', 'EST',
       'ETH', 'EU28', 'FIN', 'FJI', 'FRA', 'GAB', 'GBR', 'GEO', 'GHA',
       'GIN', 'GMB', 'GNB', 'GNQ', 'GRC', 'GRD', 'GTM', 'GUY', 'HKG',
       'HND', 'HRV', 'HTI', 'HUN', 'IDN', 'IND', 'IRL', 'IRN', 'IRQ',
       'ISL', 'ISR', 'ITA', 'JAM', 'JOR', 'JPN', 'KAZ', 'KEN', 'KGZ',
       'KHM', 'KIR', 'KNA', 'KOR', 'KWT', 'LAO', 'LBN', 'LBR', 'LBY',
       'LCA', 'LDC', 'LIE', 'LKA', 'LSO', 'LTU', 'LUX', 'LVA', 'MAC',
       'MAR', 'MCO', 'MDA', 'MDG', 'MDV', 'MEX', 'MHL', 'MKD', 'MLI',
       'ML

See the **[source](https://zenodo.org/record/4479172#data-format-description-columns)** for more details about the additional custom codes for country groups.

In [15]:
len(primap_hist_data.country.unique())

215

## Category

IPCC (Intergovernmental Panel on Climate Change) 2006 categories for emissions. Some aggregate sectors have been added to the hierarchy. These begin with the prefix IPCM instead of IPC.

In [16]:
primap_hist_data.category.unique()

array(['IPC1A', 'IPC1B1', 'IPC1B2', 'IPC1B3', 'IPC1B', 'IPC1', 'IPC2B',
       'IPC2C', 'IPC2D', 'IPC2G', 'IPC2H', 'IPC2', 'IPC3A', 'IPC4',
       'IPC5', 'IPCM0EL', 'IPCMAGELV', 'IPCMAG', 'IPC2A'], dtype=object)

See the **[source](https://zenodo.org/record/4479172#data-format-description-columns)** for more details about the category description.

## Entity

Gas categories using global warming potentials (GWP) from either Second Assessment Report (SAR) or Fourth Assessment Report (AR4).


In [17]:
primap_hist_data.entity.unique()

array(['CH4', 'CO2', 'FGASESAR4', 'FGASES', 'HFCSAR4', 'HFCS',
       'KYOTOGHGAR4', 'KYOTOGHG', 'N2O', 'NF3', 'PFCSAR4', 'PFCS', 'SF6'],
      dtype=object)

See the **[source](https://zenodo.org/record/4479172#data-format-description-columns)** for more details about the gas description.

## Unit

Unit is either Gg or GgCO2eq (CO2-equivalent according to the global warming potential used).

## Remaining columns

Years from 1850-2018 so 169 years covered.

# Missing values

This part has got the aim to identify in the dataset the variables with missing values, count them and their frequencies. Then, some further analyses will be done if missing values are found.

In [3]:
def count_missing_values_per_variable(dataset):
    return pd.DataFrame(dataset.isnull().sum(), columns=['nb_missing_values'])

In [4]:
def count_frequency_missing_values(recap_table, number_total_values):
    return recap_table.assign(freq_missing_value=lambda df:
                                  round(df.nb_missing_values/number_total_values*100, 1))

In [5]:
def keep_missing_values(table):
    return table[table.nb_missing_values > 0]

In [6]:
def create_recap_table_for_no_missing_value():
    return pd.DataFrame({'Results': ['None missing value in the dataset']})

In [7]:
def sort_descending_number_missing_values(recap_table):
    return recap_table.sort_values(by='nb_missing_values', ascending=False)

In [8]:
def identify_variables_with_missing_values(dataset):
    
    recap_table = count_missing_values_per_variable(dataset)
    recap_table = keep_missing_values(recap_table)
    nb_variables_with_missing_values = len(recap_table)

    if nb_variables_with_missing_values > 0:
        recap_table = count_frequency_missing_values(recap_table, len(dataset))
        recap_table = sort_descending_number_missing_values(recap_table)
    else:
        recap_table = create_recap_table_for_no_missing_value()
    
    return recap_table

In [19]:
list_variables_with_missing_values = identify_variables_with_missing_values(primap_hist_data)
list_variables_with_missing_values

nb_missing_values  freq_missing_value
1850                104                 0.4
1956                104                 0.4
1949                104                 0.4
1950                104                 0.4
1951                104                 0.4
...                 ...                 ...
2006                 94                 0.3
2007                 72                 0.3
2008                 70                 0.2
2009                 34                 0.1
2010                  4                 0.0

[161 rows x 2 columns]

In [15]:
list_variables_with_missing_values.index.sort_values()

Index(['1850', '1851', '1852', '1853', '1854', '1855', '1856', '1857', '1858',
       '1859',
       ...
       '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009',
       '2010'],
      dtype='object', length=161)

In [10]:
identify_variables_with_missing_values(primap_hist_data[['scenario', 'country', 'category', 'entity', 'unit']])

Results
0  None missing value in the dataset

The first variables 'scenario', 'country', 'category', 'entity' and 'unit' all contain no missing values. The missing values only concern the year variables from 1850 to 2010.

In [11]:
def count_missing_values_per_row(dataset):
    return pd.DataFrame(dataset.isnull().sum(axis=1), columns=['nb_missing_values'])

In [12]:
def identify_rows_with_missing_values(dataset):
    
    recap_table = count_missing_values_per_row(dataset)
    recap_table = keep_missing_values(recap_table)
    nb_rows_with_missing_values = len(recap_table)
    
    if nb_rows_with_missing_values > 0:
        recap_table = count_frequency_missing_values(recap_table, len(dataset.columns))
        recap_table = sort_descending_number_missing_values(recap_table)
    else:
        recap_table = create_recap_table_for_no_missing_value()
        
    return recap_table

In [18]:
list_rows_with_missing_values = identify_rows_with_missing_values(primap_hist_data)
list_rows_with_missing_values

nb_missing_values  freq_missing_value
11297                161                92.5
25637                161                92.5
25608                161                92.5
11326                161                92.5
6497                 160                92.0
...                  ...                 ...
15474                154                88.5
1286                 154                88.5
15448                154                88.5
1137                 154                88.5
1163                 154                88.5

[104 rows x 2 columns]

In [38]:
filter_rows = list_rows_with_missing_values.index
rows_with_missing_values = primap_hist_data.iloc[filter_rows, :5]

In [39]:
rows_with_missing_values.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 104 entries, 11297 to 1163
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   scenario  104 non-null    object
 1   country   104 non-null    object
 2   category  104 non-null    object
 3   entity    104 non-null    object
 4   unit      104 non-null    object
dtypes: object(5)
memory usage: 4.9+ KB


In [40]:
rows_with_missing_values.category.value_counts()

IPC1B1     48
IPC2       22
IPCM0EL    16
IPC1B       6
IPC1B2      6
IPC2C       4
IPC2B       2
Name: category, dtype: int64

The half of rows with missing values concerns the category *Solid fuels*, the two other over-represented categories are *Industrial Processes and Product Use* and *National Total excluding LULUCF*.

In [44]:
rows_with_missing_values[rows_with_missing_values.category == 'IPC1B1']

scenario country category       entity     unit
11297   HISTCR     MNE   IPC1B1          N2O       Gg
25637   HISTTP     SRB   IPC1B1          N2O       Gg
25608   HISTTP     MNE   IPC1B1          N2O       Gg
11326   HISTCR     SRB   IPC1B1          N2O       Gg
11272   HISTCR     HUN   IPC1B1          N2O       Gg
25583   HISTTP     HUN   IPC1B1          N2O       Gg
14693   HISTTP     WSM   IPC1B1          CH4       Gg
14674   HISTTP     TLS   IPC1B1          CH4       Gg
245     HISTCR     COK   IPC1B1          CH4       Gg
20539   HISTTP     COK   IPC1B1  KYOTOGHGAR4  GgCO2eq
20589   HISTTP     KIR   IPC1B1  KYOTOGHGAR4  GgCO2eq
20630   HISTTP     PLW   IPC1B1  KYOTOGHGAR4  GgCO2eq
20660   HISTTP     TLS   IPC1B1  KYOTOGHGAR4  GgCO2eq
20679   HISTTP     WSM   IPC1B1  KYOTOGHGAR4  GgCO2eq
14604   HISTTP     KIR   IPC1B1          CH4       Gg
23053   HISTTP     COK   IPC1B1     KYOTOGHG  GgCO2eq
23103   HISTTP     KIR   IPC1B1     KYOTOGHG  GgCO2eq
23144   HISTTP     PLW   IPC1B1     KYOTOGHG  GgCO2eq
23174   HISTTP     TLS   IPC1B1     KYOTOGHG  GgCO2eq
23193   HISTTP     WSM   IPC1B1     KYOTOGHG  GgCO2eq
25567   HISTTP     ESP   IPC1B1          N2O       Gg
25581   HISTTP     HRV   IPC1B1          N2O       Gg
25661   HISTTP     WSM   IPC1B1          N2O       Gg
14644   HISTTP     PLW   IPC1B1          CH4       Gg
14556   HISTTP     COK   IPC1B1          CH4       Gg
293     HISTCR     KIR   IPC1B1          CH4       Gg
11350   HISTCR     WSM   IPC1B1          N2O       Gg
333     HISTCR     PLW   IPC1B1          CH4       Gg
363     HISTCR     TLS   IPC1B1          CH4       Gg
382     HISTCR     WSM   IPC1B1          CH4       Gg
6228    HISTCR     COK   IPC1B1  KYOTOGHGAR4  GgCO2eq
6278    HISTCR     KIR   IPC1B1  KYOTOGHGAR4  GgCO2eq
6319    HISTCR     PLW   IPC1B1  KYOTOGHGAR4  GgCO2eq
6349    HISTCR     TLS   IPC1B1  KYOTOGHGAR4  GgCO2eq
6368    HISTCR     WSM   IPC1B1  KYOTOGHGAR4  GgCO2eq
8792    HISTCR     KIR   IPC1B1     KYOTOGHG  GgCO2eq
8833    HISTCR     PLW   IPC1B1     KYOTOGHG  GgCO2eq
8863    HISTCR     TLS   IPC1B1     KYOTOGHG  GgCO2eq
8882    HISTCR     WSM   IPC1B1     KYOTOGHG  GgCO2eq
11256   HISTCR     ESP   IPC1B1          N2O       Gg
11270   HISTCR     HRV   IPC1B1          N2O       Gg
8742    HISTCR     COK   IPC1B1     KYOTOGHG  GgCO2eq
11231   HISTCR     BGR   IPC1B1          N2O       Gg
25542   HISTTP     BGR   IPC1B1          N2O       Gg
11266   HISTCR     GRC   IPC1B1          N2O       Gg
25577   HISTTP     GRC   IPC1B1          N2O       Gg
302     HISTCR     LTU   IPC1B1          CH4       Gg
14613   HISTTP     LTU   IPC1B1          CH4       Gg

In [41]:
rows_with_missing_values.scenario.value_counts()

HISTCR    52
HISTTP    52
Name: scenario, dtype: int64

In [42]:
rows_with_missing_values.country.value_counts()

MLT    20
MAR    16
KAZ     8
WSM     8
COK     6
TLS     6
PLW     6
KIR     6
MRT     4
OMN     4
EST     4
HUN     2
GRC     2
SRB     2
BGR     2
LTU     2
ESP     2
MNE     2
HRV     2
Name: country, dtype: int64

The most number of rows with missing values concerns Malta and Morocco.

In [45]:
rows_with_missing_values[rows_with_missing_values.country == 'MLT']

scenario country category       entity     unit
6497    HISTCR     MLT   IPC1B2  KYOTOGHGAR4  GgCO2eq
21025   HISTTP     MLT    IPC1B  KYOTOGHGAR4  GgCO2eq
20808   HISTTP     MLT   IPC1B2  KYOTOGHGAR4  GgCO2eq
6714    HISTCR     MLT    IPC1B  KYOTOGHGAR4  GgCO2eq
9011    HISTCR     MLT   IPC1B2     KYOTOGHG  GgCO2eq
15035   HISTTP     MLT    IPC1B          CH4       Gg
23322   HISTTP     MLT   IPC1B2     KYOTOGHG  GgCO2eq
23539   HISTTP     MLT    IPC1B     KYOTOGHG  GgCO2eq
14821   HISTTP     MLT   IPC1B2          CH4       Gg
724     HISTCR     MLT    IPC1B          CH4       Gg
510     HISTCR     MLT   IPC1B2          CH4       Gg
9228    HISTCR     MLT    IPC1B     KYOTOGHG  GgCO2eq
28325   HISTTP     MLT     IPC2         PFCS  GgCO2eq
28236   HISTTP     MLT  IPCM0EL      PFCSAR4  GgCO2eq
28147   HISTTP     MLT     IPC2      PFCSAR4  GgCO2eq
14103   HISTCR     MLT  IPCM0EL         PFCS  GgCO2eq
14014   HISTCR     MLT     IPC2         PFCS  GgCO2eq
28414   HISTTP     MLT  IPCM0EL         PFCS  GgCO2eq
13836   HISTCR     MLT     IPC2      PFCSAR4  GgCO2eq
13925   HISTCR     MLT  IPCM0EL      PFCSAR4  GgCO2eq

In [43]:
rows_with_missing_values.entity.value_counts()

CH4            28
N2O            16
KYOTOGHGAR4    14
KYOTOGHG       14
PFCSAR4         8
PFCS            8
FGASES          4
HFCS            4
FGASESAR4       4
HFCSAR4         4
Name: entity, dtype: int64

In [57]:
rows_with_missing_values.pivot_table(values='scenario', index=['country', 'category', 'entity'], 
                                     aggfunc='count')

scenario
country category entity               
BGR     IPC1B1   N2O                 2
COK     IPC1B1   CH4                 2
                 KYOTOGHG            2
                 KYOTOGHGAR4         2
ESP     IPC1B1   N2O                 2
EST     IPC2     CH4                 2
        IPC2C    CH4                 2
GRC     IPC1B1   N2O                 2
HRV     IPC1B1   N2O                 2
HUN     IPC1B1   N2O                 2
KAZ     IPC2     PFCS                2
                 PFCSAR4             2
        IPCM0EL  PFCS                2
                 PFCSAR4             2
KIR     IPC1B1   CH4                 2
                 KYOTOGHG            2
                 KYOTOGHGAR4         2
LTU     IPC1B1   CH4                 2
MAR     IPC2     FGASES              2
                 FGASESAR4           2
                 HFCS                2
                 HFCSAR4             2
        IPCM0EL  FGASES              2
                 FGASESAR4           2
                 HFCS                2
                 HFCSAR4             2
MLT     IPC1B    CH4                 2
                 KYOTOGHG            2
                 KYOTOGHGAR4         2
        IPC1B2   CH4                 2
                 KYOTOGHG            2
                 KYOTOGHGAR4         2
        IPC2     PFCS                2
                 PFCSAR4             2
        IPCM0EL  PFCS                2
                 PFCSAR4             2
MNE     IPC1B1   N2O                 2
MRT     IPC2     CH4                 2
        IPC2C    CH4                 2
OMN     IPC2     CH4                 2
        IPC2B    CH4                 2
PLW     IPC1B1   CH4                 2
                 KYOTOGHG            2
                 KYOTOGHGAR4         2
SRB     IPC1B1   N2O                 2
TLS     IPC1B1   CH4                 2
                 KYOTOGHG            2
                 KYOTOGHGAR4         2
WSM     IPC1B1   CH4                 2
                 KYOTOGHG            2
                 KYOTOGHGAR4         2
                 N2O                 2

The variables are globally well filled from 99.6% to 100% of completeness. But there are rows with a low filling rate included between 7.5% and 11.5%.

These rows are mostly represented by:
- the countries Malta and Morocco
- the categories *Solid fuels*, *Industrial Processes and Product Use* and *National Total excluding LULUCF*
- the gases *Methane*, *Nitrous oxide*, *Kyoto greenhouse gases (SAR)* and *Kyoto greenhouse gases (AR4)*